Fix excluded paths for custom config formats

thepry opened the pull request:
In current master, configuration is never passed to source locator, so it uses default value:
```ruby
# /lib/reek/source/source_locator.rb
  def initialize(paths, configuration: Configuration::AppConfiguration.default)
```

`Configuration::AppConfiguration.default` finds `*.reek` files in you app directory and uses them. Which causes problems in the following case:
If your configuration file isn't located in project root, or uses different format, like `config.reek.yml`, `exclude_paths` option would not work. But if you have `*.reek` file in your app root, reek will use `exclude_paths` from there, not from specified configuration file.

In my fix configuration is passed from `Application` down to ` Source::SourceLocator`

troessner commented:
Hmmmm. But doesn't this problem **only** appear when you have and want to use 2 or more Reek configuration files?

mvz commented:
Perhaps it also happens when using Reek through its API?

mvz commented:
@thepry can you elaborate a little on how you are using reek when you encounter this problem?

mvz commented on spec/reek/cli/input_spec.rb:
```diff
@@ -35,9 +39,20 @@ def argv; end
         end
+
+        it 'should use configuration for excluded paths' do
+          expect(subject).to receive(:configuration).
```
Mocking the object under test is not recommended. I would move this test to OptionInterpreter, or use a different implementation of `#configuration` in the dummy class here, that allows to actually pass in the value in the initializer.

mvz commented on lib/reek/cli/option_interpreter.rb:
```diff
@@ -13,9 +13,10 @@ class OptionInterpreter
+      def initialize(options, configuration: Configuration::AppConfiguration.default)
         @options = options
         @argv = options.argv
+        @configuration = configuration
```
There is no test for this change.

mvz commented on lib/reek/cli/input.rb:
```diff
@@ -1,5 +1,6 @@
 # frozen_string_literal: true
 require_relative '../source/source_locator'
+require_relative '../configuration/app_configuration'
```
This line is not needed since you don't refer to the class defined there by name (it is always injected/defined by implementations of `#configuration`.

thepry commented:
@mvz i just typed in configration file name `reek.config` instead of `config.reek`. And when i called from a console `reek -c reek.config` i notices that `exclude_paths` option doesn't work.

But if i have `*.reek` file, reek wil use `exclude_paths` from there, not from config i specified with `- c` option

thepry commented:
>Hmmmm. But doesn't this problem only appear when you have and want to use 2 or more Reek configuration files?

No, it can happen if you have only 1 config file, but with not `.reek` format.

mvz commented:
I think this change makes sense and we should merge this. After that, we should refactor how Application interacts with the configuration and the option interpreter.

mvz commented:
On second thought, maybe another solution would make more sense: Move `#sources` out of `OptionInterpreter` altogether (since it doesn't actually look at any of the option arguments), and move it to `Application` instead, where `configuration` is already available.

thepry commented:
> Move #sources out of OptionInterpreter altogether (since it doesn't actually look at any of the option arguments), and move it to Application instead, where configuration is already available.

@mvz I like this one! I'll take a look in a few days, when i have time :-)

thepry commented:
@mvz what do you think about latest changes?

troessner commented on lib/reek/cli/application.rb:
```diff
@@ -48,6 +49,10 @@ def configure_app_configuration(config_file)
         options.generate_todo_list ? Command::TodoListCommand : Command::ReportCommand
       end
+
+      def argv
```
You're not using this method anywhere as far as I can see. Let's remove it.

mvz replied:
I think it's needed by Files, but I think that whole module may just have to be moved here instead.

thepry replied:
Yeah, `argv` is used in `/cli/input.rb` module.

troessner commented on lib/reek/cli/input.rb:
```diff
@@ -44,6 +44,14 @@ def sources_from_argv
+      end
+
+      def configuration
+        raise NotImplementedError
```
Also not sure about those 2 methods here. I mean, it's a module already with instance methods, so it's clear that it is used as a mixin. Raising NotImplementedError here is something I find rather confusing than helpful. I'd rather remove those.

troessner commented on spec/reek/cli/input_spec.rb:
```diff
@@ -7,10 +7,22 @@ class DummyClass
+    def configuration; end
+  end
+
+  let(:path_excluded_in_configration) do
```
typo: "configuration" ;)

troessner commented:
Don't get me wrong, I love it that you fixed this bug, I just believe that we can drop at least half of the changes to our files in /lib because they are unnecessary. 
If you address my comments and squash this bad boy I'm more than happy to merge this ;)

troessner commented on spec/reek/cli/input_spec.rb:
```diff
@@ -7,10 +7,22 @@ class DummyClass
 
     def argv; end
+
+    def configuration; end
```
Wait, thats also unnecessary, isn't it?

thepry replied:
It is necessary, since we are stubbing this method for specs.

thepry commented:
@troessner Hi. I've removed unnecessary code and squashed all changes into one commit. Plz take a look again.

Also, if `Reek::CLI::Input` module is used only in `Reek::CLI::Application` and relies on it's instance method and variables, i think we should just move all methods from `Input` to `Application`. But this refactoring is not related to bugfix PR.

troessner commented:
Looks good to me! Build failing unfortunately, but that should be fixed with #900 - waiting for the build now.
You'll then just need to rebase against master and then this one is ready to go!

troessner commented:
Merging this one since #900 is in master.
Great job, thanks!

troessner commented:
And 4.0.1 is released with the fix.
